### DISCUSSION TITLE
\prop_item:Nn vs \prop_get:NnN

### DISCUSSION THREAD

**Discussion by qinglee** [lvjr/tabularray#266] [General]
Currently, tabularray uses property lists to store some data. Also use `\prop_item:Nn` to get data from the list.

`\prop_item:Nn` is expandable and looks more natural and flexible. However, as the LaTeX3 documentation mentions, `\prop_item:Nn` is slower than the non-expandable analogue `\prop_get:NnN`. Expandability is not so important for tabularray.  I'd rather use `\prop_get:NnN` for a little bit of performance.

**Comment by lvjr** [comment]
Yes, you are right. Most of the `\prop_item:Nn`s can be replaced by `\prop_get:NnN`s in the package code.

At the beginning of the package, there are three similar functions `\__tblr_prop_item` (which calls `\prop_item`), `\__tblr_spec_item` and `\__tblr_data_item`which correspond to three different data structures respectively. This make it easy to toggle among them and test which one provides the best performance.

If we are going to add `\__tblr_prop_get` which calls `\prop_get`, then we also need to add `\__tblr_spec_get` and `\__tblr_data_get` too.

PS: I'm sometimes confused because in some cases expandable macros are faster while in other cases non-expandable macros are faster.
